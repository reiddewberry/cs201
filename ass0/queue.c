    #include <stdio.h>
    #include <stdlib.h>
    #include <assert.h>
    #include "queue.h"
    #include "dll.h"

    typedef struct queue QUEUE;

    struct QUEUE
        {
        SLL *list;
        void *(display)(void *,FILE *);
        void *(free)(void *);
        };

    QUEUE *newQUEUE(void (*d)(void *,FILE *),void (*f)(void *)){
        QUEUE *items = malloc(sizeof(QUEUE));
        assert(items != 0);
        items->list = newSLL(d,f);
        items->display = d;
        items->free = f;
        return items;
        }

    void enqueue(QUEUE *items,void *value){
        insertSLL(items->list,items->list->size,value);
        }

    void *dequeue(QUEUE *items){
        assert(items->list->size > 0);
        return removeSLL(items->list,0);
        }

    void *peekQUEUE(QUEUE *items){
        assert(items->list->size > 0);
        return getSLL(items->list,0);
        }

    int sizeQUEUE(QUEUE *items){
        return items->list->size;
        }

    void displayQUEUE(QUEUE *items,FILE *FP){
        fprintf(FP,"<")
        int i = 0;
        for(i=0; i < items->list->size; i++){
            fprintf(FP,getSLL(items->list,i));
            if(i < items->list->size-1){
                fprintf(FP,",")
                }
            }
        fprintf(FP,">")
        }

    void displayQUEUEdebug(QUEUE *items,FILE *FP){
        displaySLLdebug(items->list,FP);
        }

    void freeQUEUE(QUEUE *items){
        freeSLL(items->list);
        free(items);
        }
