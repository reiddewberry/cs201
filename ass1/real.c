#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <assert.h>
#include "real.h"

struct REAL
    {
    double value;
    };

REAL *
newREAL(double x)
    {
    REAL *p = malloc(sizeof(REAL));
    assert(p != 0);
    p->value = x;
    return p;
    }

double
getREAL(REAL *v)
    {
    return v->value;
    }

double
setREAL(REAL *v,double x)
    {
    double old = v->value;
    v->value = x;
    return old;
    }

void
displayREAL(void *v,FILE *fp)
    {
    fprintf(fp,"%lf",getREAL((REAL *) v));
    }

int
compareREAL(void *real1,void *real2)
    {
    double temp = getREAL((REAL *) real1) - getREAL((REAL *) real2);
    if(temp < 0)
        {
        return -1;
        }
    if(temp > 0)
        {
        return 1;
        }
    else{
        return 0;
        }
    }

int
compareREALdec(void *real1,void *real2)
    {
    double temp = getREAL((REAL *) real2) - getREAL((REAL *) real1);
    if(temp < 0)
        {
        return -1;
        }
    if(temp > 0)
        {
        return 1;
        }
    else{
        return 0;
        }
    }

void
freeREAL(void *v)
    {
    free((REAL *)v);
    }
