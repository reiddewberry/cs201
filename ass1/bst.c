    #include <stdio.h>
    #include <stdlib.h>
    #include <assert.h>
    #include "bst.h"
    #include "queue.h"

    typedef struct bstnode{
        void *data;
        struct bstnode *parent;
        struct bstnode *right;
        struct bstnode *left;
        }BSTNODE;

    BSTNODE *newBSTNODE(void *value){
        BSTNODE *bstNode = malloc(sizeof(BSTNODE));
        assert(bstNode != 0);
        bstNode->data = value;
        bstNode->parent = 0;
        bstNode->right = 0;
        bstNode->left = 0;
        return bstNode;
        }

    void *getBSTNODEvalue(BSTNODE *node){
        return node->data;
        }

    void setBSTNODEvalue(BSTNODE *node,void *value){
        node->data = value;
        }

    BSTNODE *getBSTNODEleft(BSTNODE *node){
        if(node->left == 0){
            return 0;
            }
        return node->left;
        }

    void setBSTNODEleft(BSTNODE *node,BSTNODE *replacement){
        node->left = replacement;
        }

    BSTNODE *getBSTNODEright(BSTNODE *node){
        if(node->right == 0){
            return 0;
            }
        return node->right;
        }

    void setBSTNODEright(BSTNODE *node,BSTNODE *replacement){
        node->right = replacement;
        }

    BSTNODE *getBSTNODEparent(BSTNODE *node){
        if(node->parent == 0){
            return 0;
            }
        return node->parent;
        }

    void setBSTNODEparent(BSTNODE *node,BSTNODE *replacement){
        node->parent = replacement;
        }

    void freeBSTNODE(BSTNODE *node,void (*freefunct)(void *value)){
        if(freefunct != 0){
            freefunct(node->data);
            }
        free(node);
        }

//Binary Search Tree section from here on

    void swapFunct(BSTNODE *move,BSTNODE *dest){
        void *valHolder = move->data;
        move->data = dest->data;
        dest->data = valHolder;
        }

    typedef struct bst{
        BSTNODE *root;
        int size;
        void (*display)(void *, FILE *);
        int (*comp)(void *, void *);
        void (*swap)(BSTNODE *move, BSTNODE *dest);
        void (*free)(void *);
        }BST;

    BST *newBST(
        void (*display)(void *value,FILE *FP),        //display
        int (*comp)(void *val1,void *val2),           //comparator. - lt, 0 eq, +  gt
        void (*swap)(BSTNODE *move,BSTNODE *dest),    //swapper
        void (*free)(void *value)){                   //free
        BST *newBST = malloc(sizeof(BST));
        assert(newBST != 0);
        newBST->root = 0;
        newBST->size = 0;
        newBST->display = display;
        newBST->comp = comp;
        if(swap == 0){
            newBST->swap = swapFunct;
            }
        else{newBST->swap = swap;}
        newBST->free = free;
        return newBST;
        };

    BSTNODE *getBSTroot(BST *tree){
        return tree->root;
        }

    void setBSTroot(BST *tree,BSTNODE *replacement){
        tree->root = replacement;
        }

    void setBSTsize(BST *tree,int size){
        tree->size = size;
        }

    BSTNODE *insertBST(BST *tree,void *value){
        BSTNODE *newNode = malloc(sizeof(BSTNODE));
        assert(newNode != 0);
        newNode->left = 0;
        newNode->right = 0;
        newNode->data = value;
        if(tree->root == 0){
            tree->root = newNode;
            tree->size += 1;
            return newNode;
            }
        BSTNODE *compNode = tree->root;
        while(1){
            if(tree->comp(newNode->data,compNode->data) < 0){
                if(compNode->left == 0){
                    compNode->left = newNode;
                    newNode->parent = compNode;
                    tree->size += 1;
                    return newNode;
                    }
                compNode = compNode->left;
                }
            else{
                if(compNode->right == 0){
                    compNode->right = newNode;
                    newNode->parent = compNode;
                    tree->size += 1;
                    return newNode;
                    }
                compNode = compNode->right;
                }
            }
        }

    BSTNODE *findBST(BST *tree,void *value){
        BSTNODE *compNode = tree->root;
        if(tree->size == 0){
            return 0;
            }
        while(1){
            if(tree->comp(value,compNode->data) < 0){
                if(compNode->left == 0){
                    return 0;
                    }
                compNode = compNode->left;
                }
            else{
                if(tree->comp(value,compNode->data) == 0){
                    return compNode;
                    }
                if(compNode->right == 0){
                    return 0;
                    }
                compNode = compNode->right;
                }
            }
        }

    BSTNODE *deleteBST(BST *tree,void *value){
        BSTNODE *delNode = findBST(tree,value);
        delNode = swapToLeafBST(tree,delNode);
        pruneLeafBST(tree,delNode);
        if(delNode != 0){
            tree->size--;
            }
        return delNode;
        }

    static BSTNODE *getSuccessor(BST *tree,BSTNODE *leafNode){
        BSTNODE *initNode = leafNode;
        while(!(leafNode->left == 0 && leafNode->right == 0)){
            if(leafNode->right != 0){
                if(leafNode->right->left != 0){
                    leafNode = leafNode->right->left;
                    while(leafNode->left != 0){
                        leafNode = leafNode->left;
                        }
                    tree->swap(initNode,leafNode);
                    initNode = leafNode;
                    }
                else{
                    leafNode =leafNode->right;
                    tree->swap(initNode,leafNode);
                    initNode = leafNode;
                    }
                }
            if(leafNode->left != 0){
                leafNode = leafNode->left;
                }
            }
        tree->swap(initNode,leafNode);
        return leafNode;
        }

    static BSTNODE *getPredecessor(BST *tree,BSTNODE *leafNode){
        BSTNODE *initNode = leafNode;
        while(!(leafNode->left == 0 && leafNode->right == 0)){
            if(leafNode->left != 0){
                if(leafNode->left->right != 0){
                    leafNode = leafNode->left->right;
                    while(leafNode->right != 0){
                        leafNode = leafNode->right;
                        }
                    tree->swap(initNode,leafNode);
                    initNode = leafNode;
                    }
                else{
                    leafNode =leafNode->left;
                    tree->swap(initNode,leafNode);
                    initNode = leafNode;
                    }
                }
            if(leafNode->right != 0){
                leafNode = leafNode->right;
                }
            }
        tree->swap(initNode,leafNode);
        return leafNode;
        }

    BSTNODE *swapToLeafBST(BST *tree,BSTNODE *node){
        BSTNODE *leafNode = node;
        if(tree->size == 0 || leafNode == 0){
            return leafNode;
            }
        if(leafNode->right != 0){
            leafNode = getSuccessor(tree,leafNode);
            }
        if(leafNode->left != 0){
            leafNode = getPredecessor(tree,leafNode);
            }
        return leafNode;
        }

    void pruneLeafBST(BST *tree,BSTNODE *leaf){
        if(leaf == 0){
            return;
            }
        if(leaf == tree->root){
            tree->root = 0;
            return;
            }
        if(leaf->parent->right == leaf){
            leaf->parent->right = 0;
            leaf->parent = 0;
            }
        else{
            leaf->parent->left = 0;
            leaf->parent = 0;
            }
        }

    int sizeBST(BST *tree){
        return tree->size;
        }

    static int maxDepthBST(BSTNODE *root){
        if(root == 0){
            return 0;
            }
        else{
            int leftDepth = maxDepthBST(root->left);
            int rightDepth = maxDepthBST(root->right);
            if(leftDepth > rightDepth){
                return leftDepth+1;
                }
            else{
                return rightDepth+1;
                }
            }
        }

    static int minDepthBST(BSTNODE *root){
        if(root->left == 0 || root->right == 0){
            return 1;
            }
        if(root->left == 0){
            return minDepthBST(root->right)+1;
            }
        if(root->right == 0){
            return minDepthBST(root->left)+1;
            }
        if(minDepthBST(root->left) <= minDepthBST(root->right)){
            return minDepthBST(root->left)+1;
            }
        else{
            return minDepthBST(root->right)+1;
            }
        }

    void statisticsBST(BST *tree,FILE *fp){
        fprintf(fp,"Nodes: %d\n", tree->size);
        if(tree->root == 0){
            fprintf(fp,"Minimum depth: -1\n");
            fprintf(fp,"Maximum depth: -1\n");
            return;
            }
        //min depth
        int min = minDepthBST(tree->root);
        fprintf(fp,"Minimum dpeth: %d\n",min);
        //max depth
        int max = maxDepthBST(tree->root);
        fprintf(fp,"Maximum depth: %d\n",max);
        }

    static void preDisplayBST(BST *tree,BSTNODE *printNode,FILE *fp){
        fprintf(fp,"[");
        if(tree->size == 0){
            return;
            }
        tree->display(printNode->data,fp);
        if(printNode->left != 0){
            fprintf(fp," ");
            preDisplayBST(tree,printNode->left,fp);
            fprintf(fp,"]");
            }

        if(printNode->right != 0){
            fprintf(fp," ");
            preDisplayBST(tree,printNode->right,fp);
            fprintf(fp,"]");
            }
        }

    void displayBST(BST *tree,FILE *fp){
        preDisplayBST(tree,tree->root,fp);
        fprintf(fp,"]");
        }

    static void levelDisplayBST(BST *tree,FILE *fp){
        QUEUE *printQ = newQUEUE(0,0);
        enqueue(printQ,tree->root);
        BSTNODE *printNode;
        while(1){
            int sizeQ = sizeQUEUE(printQ);
            if(sizeQ == 0){
                break;
                }
            while(sizeQ > 0){
                printNode = peekQUEUE(printQ);
                tree->display(printNode->data,fp);
                dequeue(printQ);
                if(printNode->left != 0){
                    enqueue(printQ,printNode->left);
                    }
                if(printNode->right != 0){
                    enqueue(printQ,printNode->right);
                    }
                sizeQ -= 1;
                if(sizeQ > 0){
                    fprintf(fp," ");
                    }
                }
            fprintf(fp,"\n");
            }
        freeQUEUE(printQ);
        }

    void displayBSTdebug(BST *tree,FILE *fp){
        if(tree->size == 0){return;}
        levelDisplayBST(tree,fp);
        }

    void freeBST(BST *tree){
        BSTNODE *remNode;
        while(tree->root != 0){
            remNode = swapToLeafBST(tree,tree->root);
            pruneLeafBST(tree,remNode);
            if(tree->free != 0){
                tree->free(remNode->data);
                }
            free(remNode);
            }
        free(tree);
        }
