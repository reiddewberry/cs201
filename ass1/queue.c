/*
Chandler Reid Dewberry
Queue Program
*/
    #include <stdio.h>
    #include <stdlib.h>
    #include <assert.h>
    #include "queue.h"
    #include "sll.h"

//QUEUE structure
    typedef struct queue
        {
        SLL *list;                      //pointer to underlying Singly-linked List struct
        void (*display)(void *,FILE *); //passed in display function
        void (*free)(void *);           //passed in free function
        }QUEUE;

/*
Public Function to create and initialize a new QUEUE
Parameters:
    *d - Display function to display whatever is stored in the QUEUE
    *f - Frees whatever is stored in the QUEUE
*/
    QUEUE *newQUEUE(void (*d)(void *,FILE *),void (*f)(void *)){
        //malloc to allocate memory and assert to verify there is enough
        QUEUE *items = malloc(sizeof(QUEUE));
        assert(items != 0);
        //create and initilize the SLL
        items->list = newSLL(d,f);
        items->display = d;
        items->free = f;
        return items;
        }

//Inserts the passed in value into the back of the SLL
    void enqueue(QUEUE *items,void *value){
        insertSLL(items->list,sizeSLL(items->list),value);
        }

//Removes and Returns what is at the front of the SSL
    void *dequeue(QUEUE *items){
        //assert to verify the list isn't empty and the program is working
        assert(sizeSLL(items->list) > 0);
        return removeSLL(items->list,0);
        }

//Doesn't Remove but Returns the pointer to what is at the front of the SLL
    void *peekQUEUE(QUEUE *items){
        assert(sizeSLL(items->list) > 0);
        return getSLL(items->list,0);
        }

//Returns the # of items in the QUEUE
    int sizeQUEUE(QUEUE *items){
        return sizeSLL(items->list);
        }

/*
Prints the Queue in order
2 Parameters:
    items - the QUEUE to be displayed
    fp - file pointer to print to
ex.
<3,5,2,7>
*/
    void displayQUEUE(QUEUE *items,FILE *FP){
        fprintf(FP,"<");
        int i = 0;
        //iterates throught the SLL by the index #
        for(i=0; i < sizeSLL(items->list); i++){
            items->display(getSLL(items->list,i),FP);
            if(i < sizeSLL(items->list)-1){
                fprintf(FP,",");
                }
            }
        fprintf(FP,">");
        }

//Prints the Queue but with more info
//ex. "head->{3,5,2,7},tail->{7}"
    void displayQUEUEdebug(QUEUE *items,FILE *FP){
        displaySLLdebug(items->list,FP);
        }

/*
Frees: 
    QUEUE
    Underlying SLL
    Data - if a Freeing function was passed in
*/
    void freeQUEUE(QUEUE *items){
        freeSLL(items->list);
        free(items);
        }
