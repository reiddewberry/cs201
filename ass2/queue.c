    #include <stdio.h>
    #include <stdlib.h>
    #include <assert.h>
    #include "queue.h"
    #include "sll.h"

    typedef struct queue
        {
        SLL *list;
        void (*display)(void *,FILE *);
        void (*free)(void *);
        }QUEUE;

    QUEUE *newQUEUE(void (*d)(void *,FILE *),void (*f)(void *)){
        QUEUE *items = malloc(sizeof(QUEUE));
        assert(items != 0);
        items->list = newSLL(d,f);
        items->display = d;
        items->free = f;
        return items;
        }

    void enqueue(QUEUE *items,void *value){
        insertSLL(items->list,sizeSLL(items->list),value);
        }

    void *dequeue(QUEUE *items){
        assert(sizeSLL(items->list) > 0);
        return removeSLL(items->list,0);
        }

    void *peekQUEUE(QUEUE *items){
        assert(sizeSLL(items->list) > 0);
        return getSLL(items->list,0);
        }

    int sizeQUEUE(QUEUE *items){
        return sizeSLL(items->list);
        }

    void displayQUEUE(QUEUE *items,FILE *FP){
        fprintf(FP,"<");
        int i = 0;
        for(i=0; i < sizeSLL(items->list); i++){
            items->display(getSLL(items->list,i),FP);
            if(i < sizeSLL(items->list)-1){
                fprintf(FP,",");
                }
            }
        fprintf(FP,">");
        }

    void displayQUEUEdebug(QUEUE *items,FILE *FP){
        displaySLLdebug(items->list,FP);
        }

    void freeQUEUE(QUEUE *items){
        freeSLL(items->list);
        free(items);
        }
