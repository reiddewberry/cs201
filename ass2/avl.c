    #include <stdio.h>
    #include <stdlib.h>
    #include <assert.h>
    #include "bst.h"
    #include "queue.h"

    typedef struct avl{
        void *data;
        }AVL;

    AVL *newAVL(
        void (*)(void *,FILE *),           //display
        int (*)(void *,void *),            //comparator
        void (*)(void *));                 //freeing function


    void insertAVL(AVL *,void *);
    int findAVLcount(AVL *,void *);
    void *findAVL(AVL *,void *);
    void *deleteAVL(AVL *,void *);
    int sizeAVL(AVL *);
    int duplicatesAVL(AVL *);
    void statisticsAVL(AVL *,FILE *);
    void displayAVL(AVL *,FILE *);
    void displayAVLdebug(AVL *,FILE *);
    void freeAVL(AVL *);
