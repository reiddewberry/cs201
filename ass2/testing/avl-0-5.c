#include <stdio.h>
#include <stdlib.h>
#include "integer.h"
#include "real.h"
#include "string.h"
#include "queue.h"
#include "bst.h"
#include "avl.h"

void srandom(unsigned int);
long int random(void);

int
main(void)
    {
    srandom(106);
    //INTEGER test of AVL, nominal size is 189
    int i;
    AVL *p = newAVL(displayINTEGER,compareINTEGER,freeINTEGER);
    for (i = 0; i < 189; ++i)
        {
        int j = random() % 189;
        INTEGER *a = newINTEGER(j);
        if (findAVL(p,a) == 0)
            insertAVL(p,a);
        else
            {
            insertAVL(p,a);
            freeINTEGER(a);
            }
        }
    if (sizeAVL(p) < 200)
        {
        printf("AVL:\n");
        displayAVL(p,stdout);
        printf("debug: ");
        displayAVLdebug(p,stdout);
        printf("\n");
        }
    printf("size is %d\n",sizeAVL(p));
    statisticsAVL(p,stdout);
    freeAVL(p);
    return 0;
    }
