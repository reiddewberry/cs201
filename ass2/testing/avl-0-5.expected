AVL:
0: 130+(130+)X
1: 63(130+)L 169[2](130+)R
2: 32(63)L 95(63)R 153[4](169[2])L 175[2]-(169[2])R
3: 20[3](32)L 41-(32)R 76-(95)L 108[2](95)R 142[2](153[4])L 158[2]-(153[4])R 172[2]-(175[2]-)L 185+(175[2]-)R
4: 11[2](20[3])L 28[2]+(20[3])R 36[2]-(41-)L 50+(41-)R 70(76-)L 87(76-)R 102(108[2])L 120[3](108[2])R 135(142[2])L 147[2](142[2])R 155[2](158[2]-)L 163[2](158[2]-)R =171[2](172[2]-)L 173[2]-(172[2]-)R 178-(185+)L 188[2]+(185+)R
5: 6[2]+(11[2])L 18+(11[2])R 23(28[2]+)L 30[2]-(28[2]+)R =33[2](36[2]-)L 40+(36[2]-)R 43-(50+)L 58(50+)R 65[2](70)L 74(70)R 79(87)L 92+(87)R 100+(102)L 106[2]+(102)R 115(120[3])L 125[2](120[3])R 131[2]-(135)L 137(135)R 144(147[2])L 151[2](147[2])R =154(155[2])L =157[3](155[2])R 160[2](163[2])L 167[2](163[2])R =174(173[2]-)R =177[2](178-)L 181+(178-)R =187(188[2]+)L
6: 3[2](6[2]+)L =10(6[2]+)R 14(18+)L =19[4](18+)R 22[2]+(23)L 27[2]+(23)R =31(30[2]-)R =38(40+)L =42(43-)L 44-(43-)R =52[2](58)L =59(58)R =64(65[2])L =67[2](65[2])R =72[2](74)L =75(74)R 78[3]+(79)L 80-(79)R 89-(92+)L =93(92+)R 98[2]-(100+)L =101(100+)R 104-(106[2]+)L =107(106[2]+)R 111(115)L 117(115)R 122[2](125[2])L 128[3](125[2])R =132(131[2]-)R =136(137)L =141(137)R =143(144)L =146[2](144)R =150[2](151[2])L =152(151[2])R =159[2](160[2])L =162[4](160[2])R =166[2](167[2])L =168(167[2])R =180(181+)L
7: =2(3[2])L =5(3[2])R =13(14)L =15(14)R =21[2](22[2]+)L =25(27[2]+)L =45[2](44-)R =77(78[3]+)L =83[3](80-)R =91(89-)R =99(98[2]-)R =105[2](104-)R =109[3](111)L =113[3](111)R =116[2](117)L =119[2](117)R =121[2](122[2])L =124(122[2])R =126[2](128[3])L =129(128[3])R
debug: [130+ [63 [32 [20[3] [11[2] [6[2]+ [3[2] [2] [5]] [10]] [18+ [14 [13] [15]] [19[4]]]] [28[2]+ [23 [22[2]+ [21[2]]] [27[2]+ [25]]] [30[2]- [31]]]] [41- [36[2]- [33[2]] [40+ [38]]] [50+ [43- [42] [44- [45[2]]]] [58 [52[2]] [59]]]]] [95 [76- [70 [65[2] [64] [67[2]]] [74 [72[2]] [75]]] [87 [79 [78[3]+ [77]] [80- [83[3]]]] [92+ [89- [91]] [93]]]] [108[2] [102 [100+ [98[2]- [99]] [101]] [106[2]+ [104- [105[2]]] [107]]] [120[3] [115 [111 [109[3]] [113[3]]] [117 [116[2]] [119[2]]]] [125[2] [122[2] [121[2]] [124]] [128[3] [126[2]] [129]]]]]]] [169[2] [153[4] [142[2] [135 [131[2]- [132]] [137 [136] [141]]] [147[2] [144 [143] [146[2]]] [151[2] [150[2]] [152]]]] [158[2]- [155[2] [154] [157[3]]] [163[2] [160[2] [159[2]] [162[4]]] [167[2] [166[2]] [168]]]]] [175[2]- [172[2]- [171[2]] [173[2]- [174]]] [185+ [178- [177[2]] [181+ [180]]] [188[2]+ [187]]]]]]
size is 119
Duplicates: 70
Nodes: 119
Minimum depth: 4
Maximum depth: 7
